File: server/src/controllers/MouseController.go
package controllers

import (
	"computer-controller-server/src/models"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/go-vgo/robotgo"
)

type MouseController struct {
}

func NewMouseController() *MouseController {
	return &MouseController{}
}

func (mc *MouseController) Binding(ginEngine *gin.Engine) {
	routerGroup := ginEngine.Group("mouse")
	routerGroup.POST("/scroll-relative", mc.scrollRelative)
	routerGroup.GET("/click-left", mc.clickLeft)
	routerGroup.GET("/double-click-left", mc.doubleClickLeft)
	routerGroup.GET("/click-right", mc.clickRight)
	routerGroup.POST("/move-relative", mc.moveRelative)
}

func (mc *MouseController) scrollRelative(ctx *gin.Context) {
	var position models.Position
	if err := ctx.ShouldBindJSON(&position); err != nil {
		ctx.AbortWithError(http.StatusBadRequest, err)
		return
	}
	robotgo.Scroll(position.X, position.Y)
	ctx.IndentedJSON(http.StatusOK, &gin.H{})
}

func (mc *MouseController) clickLeft(ctx *gin.Context) {
	robotgo.Click(robotgo.Left)
	ctx.IndentedJSON(http.StatusOK, &gin.H{})
}

func (mc *MouseController) doubleClickLeft(ctx *gin.Context) {
	robotgo.Click(robotgo.Left, true)
	ctx.IndentedJSON(http.StatusOK, &gin.H{})
}

func (mc *MouseController) clickRight(ctx *gin.Context) {
	robotgo.Click(robotgo.Right)
	ctx.IndentedJSON(http.StatusOK, &gin.H{})
}

func (mc *MouseController) moveRelative(ctx *gin.Context) {
	var position models.Position
	if err := ctx.ShouldBindJSON(&position); err != nil {
		ctx.AbortWithError(http.StatusBadRequest, err)
		return
	}
	robotgo.MoveRelative(position.X, position.Y)
	ctx.IndentedJSON(http.StatusOK, &gin.H{})
}

File: server/client/src/apis/MouseController.ts
import { Position } from "nipplejs";

class MouseController {
  private static readonly CONTROLLER_NAME: string = "mouse";

  /**
   * 添加偏移
   * @param x x 軸偏移量
   * @param y y 軸偏移量
   */
  public scrollRelative(x: number, y: number): Promise<Response> {
    const url = `${MouseController.CONTROLLER_NAME}/scroll-relative`;
    const body: Position = {
      x: Math.round(x),
      y: Math.round(y),
    };
    if (body.x === 0 && body.y === 0) {
      return new Promise((resolve) => resolve(new Response()));
    }
    return fetch(url, {
      body: JSON.stringify(body),
      method: "POST",
    });
  }

  public clickRight(): Promise<Response> {
    const url = `${MouseController.CONTROLLER_NAME}/click-right`;
    return fetch(url);
  }

  public clickLeft(): Promise<Response> {
    const url = `${MouseController.CONTROLLER_NAME}/click-left`;
    return fetch(url);
  }

  public doubleClickLeft(): Promise<Response> {
    const url = `${MouseController.CONTROLLER_NAME}/double-click-left`;
    return fetch(url);
  }

  /**
   * 添加偏移
   * @param x x 軸偏移量
   * @param y y 軸偏移量
   */
  public moveRelative(x: number, y: number): Promise<Response> {
    const url = `${MouseController.CONTROLLER_NAME}/move-relative`;
    const body: Position = {
      x: Math.round(x),
      y: Math.round(y),
    };
    if (body.x === 0 && body.y === 0) {
      return new Promise((resolve) => resolve(new Response()));
    }
    return fetch(url, {
      body: JSON.stringify(body),
      method: "POST",
    });
  }
}
export const mouseController = new MouseController();
